### ISSUE TITLE
Questions re html layout, classes, etc.

### ISSUE THREAD

**Issue by MSP-Greg** [lsegal/yard#960]
First, sorry for the stupid question(s) Sunday.  Helping a golf fanatic (watching Masters) with taxes.  I keep looking for YARD info in yard-js...

Anyway, wondering about some things in the YARD html.  I'm more or less done with the code for yard-t2, and I was looking at changes to make the 'content to markup' ratio a little higher, but it may involve code outside the 'template system.'  Also, you're probably a lot more familiar with what people have done as far as other templates / plugins, and their assumptions about markup.

Hence, would pull requests for the following be okay (don't have the code done yet)?

1. Summary Section(s) Items - Many of the li elements have a class 'public'.  Can't find that word anywhere in the css or script files.  See templates/default/module/html/item_summary.erb, line 1.

2. Anchors - many are surrounded by a span element with a class of 'object_link'.  css and script files have the class in them, but it is always in a selector paired with a, like '.object_link a'.  Hence, could the element be removed?  See YARD::Templates::Helpers::HtmlHelper#link_object, line 278.

There's more, thought I'd just start with two.  Having spent a bit of time with a 'Method List' of 20k items (AWS Ruby SDK), anything to remove elements, classes, etc helps with both rendering time and retrieval time.

Thanks,

Greg

**Comment by lsegal** [comment]
> Summary Section(s) Items - Many of the li elements have a class 'public'. Can't find that word anywhere in the css or script files. See templates/default/module/html/item_summary.erb, line 1.

The visibility of an object is passed into the className of each li element and other locations for styling. It might not be used in the default template, but the idea is that you *could* customize YARD's template/stylesheets to style based on object visibility-- like, for example, organizing all public/protected/private methods into separate searchable lists, or creating a custom toggle button that shows/hides objects of a given visibility. This information is in the template for semantics, basically.

> Anchors - many are surrounded by a span element with a class of 'object_link'. css and script files have the class in them, but it is always in a selector paired with a, like '.object_link a'. Hence, could the element be removed? See YARD::Templates::Helpers::HtmlHelper#link_object, line 278.

While technically it could, this would be a breaking change to the templates with [a pretty wide blast radius](https://github.com/search?l=css&q=object_link&type=Code&utf8=%E2%9C%93), and I would be concerned about such a change.

> There's more, thought I'd just start with two.

In general I would recommend creating incremental PRs with individual suggestions, see below for why. Note that backwards compatibility will always be a concern-- if the change knowingly affects a large set of users, I'm unlikely to accept it unless the benefits strongly outweigh the breakage (like, say, a *significant* performance boost).

> Having spent a bit of time with a 'Method List' of 20k items (AWS Ruby SDK), anything to remove elements, classes, etc helps with both rendering time and retrieval time.

If you can show (in each PR) the actual rendering time improvements per element/class removal, we can look at optimizing based on actual results. My hypothesis is that the extra classes and elements don't provide significant overhead in the general case, but if you can show examples of page elements that contribute to render time disproportionately, we can fix those on a case-by-case basis.

**Comment by MSP-Greg** [comment]
'pretty wide blast radius' - didn't know you could search GitHub like that.  Thanks.

Seems most of those are from a couple of gems, but thread_safe's had 40M downloads.  The change wouldn't break any code, just styling.  But it would break/change.

Understand compatibility issues.  In yard-t2 I'll add visibility as a data- attribute, so available in code, but not css.

As you know, re rendering time, for many 'normal' modules/classes, it's a matter of one blink of the eye or two.  I was looking at the AWS JS 'class' EC2 that's about 2 meg in size (as html), with **a lot of methods.**  Not a normal use case.  Like so big that the browser chokes on list resizing because of all the reflow in the main doc. And that's a PC, haven't tried it on a tablet...

Re PR's, going forward, I'll never submit one without discussion.

Thanks for the reply.

**Comment by lsegal** [comment]
> I was looking at the AWS JS 'class' EC2 that's about 2 meg in size (as html), with a lot of methods.

I understand the edge cases-- I guess my point is: does removing a single element, like the object_link span, reduce the render time that significantly on such a page? Or, perhaps in the same line of questioning: are there equivalent optimizations that could be applied that have the same goal but without changing structure? For example, if reducing file size does indeed have a significant impact, then there are better ways to do this-- removing extraneous whitespace would be one.

**Comment by MSP-Greg** [comment]
>does removing a single element, like the object_link span, reduce the render time that significantly on such a page?

I doubt it with just one, but for instance the default/docstring/html/index.erb template is two nested divs, both with classes --
```
<div class="docstring">
  <div class="discussion">
    <%= yieldall %>
  </div>
</div>
```
I doubt one change would make a difference, but several together might, with an emphasis on 'might'.  And maybe only for edge cases where there's a lot of methods/properties, etc.

I've worked with the extraneous whitespace issue before; tricky when white space is part of the content in pre/code elements, even worse when they're highlighted.  Obviously, can't do it globally.  Have you dealt with it before, or, any suggestions as to how best to implement?

I reduced the element count in my lists, and it did make a rendering difference, but was only apparent with large lists, like AWS or Rails.

Lastly, re operations that globally hide/show elements, tonite I just changed the last remaining ones from using style.display to css, and some operations were twice as fast.  Then again, one blink vs two.  Haven't uploaded to GitHub.io yet.

Never asked, what do you think of using XHR in YARD?

**Comment by lsegal** [comment]
> Never asked, what do you think of using XHR in YARD?

I've looked into it before. I wanted to create a modern template based on Angular/Ember/etc for a long time, but the lack of statically rendered content has historically made it a non-starter for SEO and such. Admittedly, I don't know what advancement these frameworks / search engines have made in the last little while, but server-side rendering is simply not feasible for this project, at least not in the default template.

Assuming these constraints are blocking, I wouldn't be opposed to a separate template with a more modern XHR style approach-- it would work very well for local docs or coupled with the YARD server.

Exporting the Registry to a json file, or set of json files, would be the first step to making that happen. I've also experimented with doing this before, and that's a whole other technical challenge.

**Comment by MSP-Greg** [comment]
When I mentioned XHR, I didn't really mean JSON or data oriented XHR.

What I did with yard-t2 was use XHR for the following:

* Once a YARD doc is loaded, all script & css are in the shell, and aren't loaded again
* Lists (and search results) remain when new docs are loaded.
* When loading another doc, the XHR removes everything but the body element, then pulls out the content and the 'menu', processes them, then adds to the DOM.
* Content added to the main doc via script is added before the doc content is added to the DOM.
* List search data is gathered before the list info is added to the DOM.

I've updated a lot of things in the script since I sent you the link a while ago, much from working with the edge cases like Rails and AWS Ruby / JS.  Got a lot better with Chrome timeline...

It works with both YARD server & YARD doc.

I've changed a lot of things, so it will break other plugins, but it was relatively easy to get it working with AWS.

That link again --

[http://msp-greg.github.io/](http://msp-greg.github.io/)

Also, note the app.js library, which is the script file, doc'd with yard-js, but I used a different JS parser (RKelly).

**Comment by lsegal** [comment]
I'm closing out old issues. It doesn't seem like there are any other action items / open questions here. Feel free to re-open if I hadn't responded to something. The work on yard-t2 certainly looking interesting.

**Comment by MSP-Greg** [comment]
I think we're fine.  Thanks for the help.  Very recently I uploaded new libraries to GitHub.io, four versions of Ruby and two versions of Rails.  If you get a minute, have a look; I did a few things to the C parser, changed the summary method signatures to pick up overload info, etc.

I've been trying to think of what things I've changed that might be added back into YARD, but a lot of it is integrated with other changes I've made.  I've also changed many of the ERB templates to make use of the '-' option...

I'm looking at the AWS-SDK now, trying to see how they have so many method duplicates, etc

Greg
